### PR TITLE
[libc++] Put _LIBCPP_NODEBUG on allocator trait aliases.

### PR DESCRIPTION
**Pull request by cmtice**
Put _LIBCPP_NODEBUG on the new allocator trait aliases introduced in https://github.com/llvm/llvm-project/pull/115654.  This prevents a large increase in the gdb_index size that was introduced by that PR.

### CODE REVIEW

File: libcxx/include/__memory/allocator_traits.h [hunk 1]
```diff
@@ -44,10 +44,10 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 
 // __pointer
 template <class _Tp>
-using __pointer_member = typename _Tp::pointer;
+using __pointer_member _LIBCPP_NODEBUG = typename _Tp::pointer;
 
 template <class _Tp, class _Alloc>
-using __pointer = __detected_or_t<_Tp*, __pointer_member, __libcpp_remove_reference_t<_Alloc> >;
+using __pointer _LIBCPP_NODEBUG = __detected_or_t<_Tp *, __pointer_member, __libcpp_remove_reference_t<_Alloc> >;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:50]: Please make sure to run this through clang-format!

File: libcxx/include/__memory/allocator_traits.h [hunk 2]
```diff
@@ -73,7 +73,7 @@ struct __void_pointer {
 template <class _Ptr, class _Alloc>
 struct __void_pointer<_Ptr, _Alloc, false> {
 #ifdef _LIBCPP_CXX03_LANG
-  using type _LIBCPP_NODEBUG = typename pointer_traits<_Ptr>::template rebind<void>::other;
+  using type = typename pointer_traits<_Ptr>::template rebind<void>::other;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:76]: Why removed here?

File: libcxx/include/__memory/allocator_traits.h [hunk 3]
```diff
@@ -88,18 +88,18 @@ struct __const_void_pointer {
 template <class _Ptr, class _Alloc>
 struct __const_void_pointer<_Ptr, _Alloc, false> {
 #ifdef _LIBCPP_CXX03_LANG
-  using type _LIBCPP_NODEBUG = typename pointer_traits<_Ptr>::template rebind<const void>::other;
+  using type = typename pointer_traits<_Ptr>::template rebind<const void>::other;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:91]: Why removed here?

File: libcxx/include/__memory/allocator_traits.h [hunk 4]
```diff
@@ -232,18 +230,18 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
+  using const_pointer _LIBCPP_NODEBUG                          = typename __const_pointer<value_type, pointer, allocator_type>::type;
+  using void_pointer _LIBCPP_NODEBUG                           = typename __void_pointer<pointer, allocator_type>::type;
+  using const_void_pointer _LIBCPP_NODEBUG                     = typename __const_void_pointer<pointer, allocator_type>::type;
+  using difference_type _LIBCPP_NODEBUG                        = typename __alloc_traits_difference_type<allocator_type, pointer>::type;
+  using size_type _LIBCPP_NODEBUG                              = __size_type<allocator_type, difference_type>;
+  using propagate_on_container_copy_assignment _LIBCPP_NODEBUG = __propagate_on_container_copy_assignment<allocator_type>;
+  using propagate_on_container_move_assignment _LIBCPP_NODEBUG = __propagate_on_container_move_assignment<allocator_type>;
+  using propagate_on_container_swap _LIBCPP_NODEBUG            = __propagate_on_container_swap<allocator_type>;
+  using is_always_equal _LIBCPP_NODEBUG                        = __is_always_equal<allocator_type>;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:244]: Do we really want to add `_LIBCPP_NODEBUG` to these types? What is the impact to debugging experience?

File: libcxx/include/__memory/allocator_traits.h [hunk 5]
```diff
@@ -253,11 +251,11 @@ struct _LIBCPP_TEMPLATE_VIS allocator_traits {
 #else  // _LIBCPP_CXX03_LANG
   template <class _Tp>
   struct rebind_alloc {
-    using other = __allocator_traits_rebind_t<allocator_type, _Tp>;
+    using other _LIBCPP_NODEBUG = __allocator_traits_rebind_t<allocator_type, _Tp>;
   };
   template <class _Tp>
   struct rebind_traits {
-    using other = allocator_traits<typename rebind_alloc<_Tp>::other>;
+    using other _LIBCPP_NODEBUG = allocator_traits<typename rebind_alloc<_Tp>::other>;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:258]: Same question here.

File: libcxx/include/__memory/unique_ptr.h [hunk 6]
```diff
@@ -410,7 +410,7 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr<_Tp[], _Dp>
 public:
   typedef _Tp element_type;
   typedef _Dp deleter_type;
-  using pointer = __pointer<_Tp, deleter_type>;
+  using pointer _LIBCPP_NODEBUG = __pointer<_Tp, deleter_type>;
```
Comment (ldionne) [inline libcxx/include/__memory/unique_ptr.h:413]: Same here.

**Review by ldionne — COMMENTED** [review]
I think adding `_LIBCPP_NODEBUG` makes sense on internal implementation-detail helpers, but I have questions about doing it on public API types.

File: libcxx/include/__memory/allocator_traits.h (hunk 2 above)
Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h:76]: I noticed that the stage 1 generic-cxx03 builder failed, so I thought maybe Cxx03 didn't recognize _LIBCPP_NODEBUG?

File: libcxx/include/__memory/allocator_traits.h (hunk 3 above)
Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h:91]: Same reason

File: libcxx/include/__memory/unique_ptr.h (hunk 6 above)
Comment (cmtice) [reply libcxx/include/__memory/unique_ptr.h:413]: "Same" as which comment?

File: libcxx/include/__memory/allocator_traits.h (hunk 4 above)
Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h:244]: Summary of complete findings, with respect to changes in .gdb_index size, on my test program.

Before https://github.com/llvm/llvm-project/pull/115654,  .gdb-index is  230.8 MB   (baseline)
After https://github.com/llvm/llvm-project/pull/115654:      .gdb_index is 252.4 MB   (9.4% increase)
With this complete PR:                                                      .gdb_index is 229.2 MB   (0.7% decrease!)
This PR, removing _LIBCPP_NODEBUG on lines above: .gdb_index is 229.2 MB (slight size increase of ~15KB)
Also removing it in two places questioned below:              No change to previous size.

So, in summary: Having the annotation on the lines above makes .gdb_index (for my one test case) ~ 15 KB smaller. Having the annotation on line 258 in allocator_traits.h and on line 413 in unique_ptr.h does not appear to affect the size of .gdb_index (for my one test case).

Please let me know for sure what you would like me to do.

File: libcxx/include/__memory/allocator_traits.h (hunk 2 above)
Comment (ldionne) [reply libcxx/include/__memory/allocator_traits.h:76]: I think it must have failed due to `>>`

File: libcxx/include/__memory/allocator_traits.h (hunk 4 above)
Comment (ldionne) [reply libcxx/include/__memory/allocator_traits.h]: Thanks for the details! In that case, I would apply `_LIBCPP_NODEBUG` only on the internal helpers, not on the public API typedefs. That's what we already have precedent for so that change would be 100% uncontroversial.

File: libcxx/include/__memory/unique_ptr.h (hunk 6 above)
Comment (ldionne) [reply libcxx/include/__memory/unique_ptr.h:413]: About applying to public API typedefs. I think we resolved that question above though, you can drop `_LIBCPP_NODEBUG` from all public API typedefs.

Comment (ldionne) [inline libcxx/include/__memory/unique_ptr.h:413]:
```suggestion
  using pointer = __pointer<_Tp, deleter_type>;
```

File: libcxx/include/__memory/allocator_traits.h [hunk 7]
```diff
@@ -232,32 +236,34 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:266]:
```suggestion
    using other = allocator_traits<typename rebind_alloc<_Tp>::other>;
```

File: libcxx/include/__memory/allocator_traits.h [hunk 8]
```diff
@@ -232,32 +236,34 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
-  using rebind_alloc = __allocator_traits_rebind_t<allocator_type, _Tp>;
+  using rebind_alloc _LIBCPP_NODEBUG = __allocator_traits_rebind_t<allocator_type, _Tp>;
   template <class _Tp>
-  using rebind_traits = allocator_traits<rebind_alloc<_Tp> >;
+  using rebind_traits _LIBCPP_NODEBUG = allocator_traits<rebind_alloc<_Tp> >;
 #else  // _LIBCPP_CXX03_LANG
   template <class _Tp>
   struct rebind_alloc {
-    using other = __allocator_traits_rebind_t<allocator_type, _Tp>;
+    using other _LIBCPP_NODEBUG = __allocator_traits_rebind_t<allocator_type, _Tp>;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:262]:
```suggestion
    using other = __allocator_traits_rebind_t<allocator_type, _Tp>;
```

File: libcxx/include/__memory/allocator_traits.h [hunk 9]
```diff
@@ -232,32 +236,34 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
+      __propagate_on_container_copy_assignment<allocator_type>;
+  using propagate_on_container_move_assignment _LIBCPP_NODEBUG =
+      __propagate_on_container_move_assignment<allocator_type>;
+  using propagate_on_container_swap _LIBCPP_NODEBUG = __propagate_on_container_swap<allocator_type>;
+  using is_always_equal _LIBCPP_NODEBUG             = __is_always_equal<allocator_type>;
 
 #ifndef _LIBCPP_CXX03_LANG
   template <class _Tp>
-  using rebind_alloc = __allocator_traits_rebind_t<allocator_type, _Tp>;
+  using rebind_alloc _LIBCPP_NODEBUG = __allocator_traits_rebind_t<allocator_type, _Tp>;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:256]:
```suggestion
  using rebind_alloc = __allocator_traits_rebind_t<allocator_type, _Tp>;
```

File: libcxx/include/__memory/allocator_traits.h [hunk 10]
```diff
@@ -232,32 +236,34 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
   template <class _Tp>
-  using rebind_traits = allocator_traits<rebind_alloc<_Tp> >;
+  using rebind_traits _LIBCPP_NODEBUG = allocator_traits<rebind_alloc<_Tp> >;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:258]:
```suggestion
  using rebind_traits = allocator_traits<rebind_alloc<_Tp> >;
```

File: libcxx/include/__memory/allocator_traits.h [hunk 11]
```diff
@@ -232,32 +236,34 @@ _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(allocation_result);
...
+  using void_pointer _LIBCPP_NODEBUG       = typename __void_pointer<pointer, allocator_type>::type;
+  using const_void_pointer _LIBCPP_NODEBUG = typename __const_void_pointer<pointer, allocator_type>::type;
+  using difference_type _LIBCPP_NODEBUG    = typename __alloc_traits_difference_type<allocator_type, pointer>::type;
+  using size_type _LIBCPP_NODEBUG          = __size_type<allocator_type, difference_type>;
+  using propagate_on_container_copy_assignment _LIBCPP_NODEBUG =
+      __propagate_on_container_copy_assignment<allocator_type>;
+  using propagate_on_container_move_assignment _LIBCPP_NODEBUG =
+      __propagate_on_container_move_assignment<allocator_type>;
+  using propagate_on_container_swap _LIBCPP_NODEBUG = __propagate_on_container_swap<allocator_type>;
+  using is_always_equal _LIBCPP_NODEBUG             = __is_always_equal<allocator_type>;
```
Comment (ldionne) [inline libcxx/include/__memory/allocator_traits.h:252]:
```suggestion
  using allocator_type    = _Alloc;
  using value_type          = typename allocator_type::value_type;
  using pointer             = __pointer<value_type, allocator_type>;
  using const_pointer      = typename __const_pointer<value_type, pointer, allocator_type>::type;
  using void_pointer       = typename __void_pointer<pointer, allocator_type>::type;
  using const_void_pointer = typename __const_void_pointer<pointer, allocator_type>::type;
  using difference_type    = typename __alloc_traits_difference_type<allocator_type, pointer>::type;
  using size_type          = __size_type<allocator_type, difference_type>;
  using propagate_on_container_copy_assignment =
      __propagate_on_container_copy_assignment<allocator_type>;
  using propagate_on_container_move_assignment =
      __propagate_on_container_move_assignment<allocator_type>;
  using propagate_on_container_swap = __propagate_on_container_swap<allocator_type>;
  using is_always_equal             = __is_always_equal<allocator_type>;
```

**Review by ldionne — APPROVED** [review]
LGTM with my suggestions applied + clang-format!

File: libcxx/include/__memory/allocator_traits.h (hunk 2 above)
Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h]: OK, I've restored _LIBCPP_NODEBUG to all the CXX03 aliases.

File: libcxx/include/__memory/allocator_traits.h (hunk 4 above)
Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h]: I'm not sure how to distinguish the public API typedefs from the internal helpers? What should I look for?

Comment (cmtice) [reply libcxx/include/__memory/allocator_traits.h]: Oh, that would be the changes you suggested below? Will do!

**Review by ldionne — APPROVED** [review]
This looks great now. Thanks for the patch!

This can be merged once CI is green (which it should be unless something unexpected happens).

**Comment by cmtice** [comment]
> This looks great now. Thanks for the patch!
> 
> This can be merged once CI is green (which it should be unless something unexpected happens).

All the stage 2 builds failed, but the failures don't seem to have anything to do with this change (the errors all seem to be about not finding or connecting to Docker).  Can I just ignore those failures or is there something I need to try to do about them?

**Comment by ldionne** [comment]
The CI runs on preemptable VMs, the jobs sometimes fail spuriously but they get restarted automatically.

**Review by ldionne — APPROVED** [review]
